### ISSUE TITLE
Center of mass force

### ISSUE THREAD

**Issue by peastman** [pandegroup/openmm#1107]
I'm looking into creating a new custom force to address a class of features that get requested every now and then.  The most common form of the request is for a harmonic restraint between the centers of mass of two groups of atoms.  I've also seen a request for a restraint between the center of mass of a single group of atoms and a fixed position.  Other variations include a non-harmonic force, or something based on the centroid rather than center of mass.  Then there was the force @jlmaccal requested a few years ago (and eventually wrote a plugin for) based on the mean of r^-6 for all distances between two groups of atoms.  That case is rather different in that it starts by computing an average over two particle terms rather than a set of averages over one particle terms.  But maybe we could create something general enough to handle even that?

Anyway, I'm looking for suggestions about the design of this force.  One option would be something like a CustomCompoundBondForce, but you would also define groups of atoms, and each "bond" would be computed based on the center of mass positions of groups, rather than the positions of individual atoms.  But I can think of other possible designs too, and I'm not really sure what other sorts of interactions people might want to use this for.

**Comment by jchodera** [comment]
Tagging @luirink who has also requested center-of-mass restraints for use in restraining the center of mass of a group of ligand atoms to the center of mass of a group of protein atoms lining a binding pocket.  The `CustomCompoundBondForce` you describe would be suitable there.

**Comment by luirink** [comment]
In fact I am looking to restraint the center of mass of ligand atoms to the centroid of a group of receptor atoms. What we were now working on is a virtual site based on more than 3 atoms. @kmvisscher

**Comment by peastman** [comment]
Does anyone else have thoughts on this?  Here are some specific questions I'd like feedback on.

1. Clearly the force must be able to depend on the position of a single cluster of atoms, and on the distance between two clusters.  Do we need anything beyond that?  For example, is there a use for angle and torsion terms like CustomCompoundBondForce has?  Or is that just an unnecessary complication?

2. Is center of mass the only value we care about for a cluster?  Or should we also support other types of centers?

3 Are there other important types of interactions between two clusters, that aren't a simple function of the positions of their centers, that we should also try to support?

**Comment by jchodera** [comment]
@davidlmobley: Do you have any thoughts about restraints for binding free energy calculations?

**Comment by jchodera** [comment]
My own thoughts:
1. I think distance dependence alone will be hugely useful, though I can see utility in using the powerful generality of `CustomCompoundBondForce` combined with atom cluster definitions.
2. I would suggest both "center of mass" and "center of geometry", and potentially even user-specified weights if these could be easily supported.
3. Minimum and maximum intercluster distance could be very useful.

**Comment by swails** [comment]
Angles and torsions could be useful for doing stuff like manipulating the relative orientations of different domains that are linked together. 

I think a COM-based CustomCompundBondForce would be a great start. Although other directionality type stuff (like pulling away from the plane of a surface) could also be useful.

**Comment by peastman** [comment]
> I would suggest both "center of mass" and "center of geometry"

By "center of geometry", do you mean an unweighted average?  Or (max-min)/2?

> Minimum and maximum intercluster distance could be very useful.

Could you explain what you mean be that?

**Comment by pren** [comment]
I think John is talking about flat well restraint. If the actual distance is between the lower and upper bounds, there is no forces other wise harmonic force is applied. Tinker does this too.

**Comment by jchodera** [comment]
> By "center of geometry", do you mean an unweighted average? Or
(max-min)/2?

Uniformly weighted average.

>> Minimum and maximum intercluster distance could be very useful.
>
> Could you explain what you mean be that?

Suppose we have a ligand and want to prevent it from drifting too far from
the surface of the protein.  If we define groups of atoms for ligand and
protein, a restraint based on the minimum distance between the two atomsets
could ensure this.

This might work great for Monte Carlo but there may be force
differentiability problems for MD, so a soft form of this min and max would
also be nice (unnormalized softmin and softmax with user specified
fuzziness parameter).

**Comment by jlmaccal** [comment]
It’s very useful to be able to specify a subset of dimensions. For example, to constrain only the z-component of the distance. Or, to use cylindrical restraints, so that the x,y distance matters by z does not. Typically, these would be mass weighted, but a uniform weighting, i.e. center of geometry, could be useful as well. Allowing arbitrary weights is the most general, but it would be great to make it easy for these two common use cases.

Angles and torsions might be useful, but they require a fair bit of care when used as a reaction coordinate due to Jacobians.

The r^-6 weighted restraints would be interesting, but they don’t really fit into this framework. They require computing all pairwise distances between two groups and then reducing that to a distance. The intercluser distances that @jchodera is mentioning would fall into this class as well. This type of restraint and the COM restraint have fundamentally different scaling, nm and n+m, respectively, where n and m are the sizes of the two groups.

I would be +1 for a COM or COG-based bonded force as a first step, where the positions of the two centers of mass are available as variables in the expression, which would allow for selecting out just the z-component, etc.

**Comment by jlmaccal** [comment]
@jchodera, I think the r^-6 averaged restraints typically used in NMR are the best way to do this. The distance computed is very heavily biased towards the minimum distance, while still remaining continuous. This is what e.g. HADDOCK uses to enforce distances between domains.

**Comment by rmcgibbo** [comment]
I think flat-bottomed harmonic potentials are pretty common too. e.g. http://onlinelibrary.wiley.com/doi/10.1002/pro.2770/abstract from last week (although this isn't with groups of atoms, just between alpha carbons).

**Comment by peastman** [comment]
Flat bottomed potential is easy to do.  You just replace "x^2" with "max(0, x-a)^2", where a is the radius of the flat region.

**Comment by jlmaccal** [comment]
@rmcgibbo I think there are two different issues. First, how is the "distance" between two groups defined? Second, what is the potential given that distance? I think the aim would be to have a custom force that supports pretty much any potential, like the existing forces do, but where the calculation of the "distance" is already baked into the force.

**Comment by peastman** [comment]
Here is a candidate proposal:

The API will be mostly identical to CustomCompoundBondForce, except everywhere that refers to a particle, replace it by the center of a group of particles.  There will be an `addParticleGroup()` method to define what the groups are.  That method will take a list of particle indices to define the particles in that group, and optionally a list of weights to use.  If the second argument is omitted, it will use particle masses as weights.

You will then define bonds as usual.  The first argument to `addBond()` will be a list of group indices, not particle indices.

Does that sound reasonable?

**Comment by jlmaccal** [comment]
:+1:

**Comment by jchodera** [comment]
> @jchodera, I think the r^-6 averaged restraints typically used in NMR are
the best way to do this. The distance computed is very heavily biased
towards the minimum distance, while still remaining continuous. This is
what e.g. HADDOCK uses to enforce distances between domains.

Well, it's certainly *a* way to do it, but I would want to tune the
softness. I like my exponential based log(sum(exp(k*r_I)))/k schemes...

**Comment by jchodera** [comment]
> It’s very useful to be able to specify a subset of dimensions.

+1

> The r^-6 weighted restraints would be interesting, but they don’t really
fit into this framework. They require computing all pairwise distances
between two groups and then reducing that to a distance. The intercluser
distances that @jchodera is mentioning would fall into this class as well.
This type of restraint and the COM restraint have fundamentally different
scaling, nm and n+m, respectively, where n and m are the sizes of the two
groups.

Agree that the scaling is different, but it is hard to imagine cases where
this would be more expensive than NonbondedForce or if so, if that would
really be a problem.

A generalization of that idea would be a force that would
* for each specified pair of atom groups
  * map a function onto each pair of atoms in each group
  * reduce these distances according to some reduction operation
  * compute an energy as a function of that reduced value

> I would be +1 for a COM or COG-based bonded force as a first step, where
the positions of the two centers of mass are available as variables in the
expression, which would allow for selecting out just the z-component, etc.

It could definitely be useful to have both r and x1, x2, y1, y2, etc.

**Comment by davidlmobley** [comment]
Thanks, @jchodera , for tagging me here. @nividic will also be interested, but may not have comments at this point. We've been very interested in having this for binding free energy calculations with Yank and it has not been obvious how to do this. 

>Do you have any thoughts about restraints for binding free energy calculations?

I have thoughts, though not conclusions. Particularly, we really want to be able to do flexible restraints. Our *thought* is that at least flat bottom restraints will be needed for maximal efficiency in some types of binding free energy calculations. But we would also like to be able to apply other types of restraints, such as pure harmonic restraints (which I suppose are a less general version of flat bottom) and even angle and torsional restraints. Basically no one has yet looked at the efficiency of binding free energy calculations as a function of the type of restraints used (though @nividic and I are thinking about such a study) so the verdict is still out. The type we have the most experience with using are using six restraints between the protein and ligand - one bond, two angles, and three torsions. 

At @peastman -- 
>... angle and torsion terms like CustomCompoundBondForce has? Or is that just an unnecessary complication?

*We* would very much like to be able to use angles and torsions. The science is still unclear as to whether these are the most efficient approach to restraints, but we think that in some cases they are going to be a good idea, and we would very much like to have the ability to at least test them. Otherwise we will be forced to use other codes which DO implement these restraints for testing them out, which would be a bit annoying since we're hoping to shift more of our work in this direction.

**Comment by jlmaccal** [comment]
> Agree that the scaling is different, but it is hard to imagine cases where this would be more expensive than NonbondedForce or if so, if that would really be a problem.

@jchodera, sorry I wasn't clear. I don't think it's too expensive at all. What I meant is that these require different algorithms and thus ought to be implemented as different custom forces.

> A generalization of that idea would be a force that would
>  * for each specified pair of atom groups
>  * map a function onto each pair of atoms in each group
>  * reduce these distances according to some reduction operation
>  * compute an energy as a function of that reduced value

At first glance, I like this idea. We would need to think a bit about the mechanics though, especially for the reduction operation, which doesn't have a clear counterpart in any of the existing custom forces. I've created another issue to discuss this type of force, leaving this thread for COM-like forces.

**Comment by peastman** [comment]
@davidlmobley Just to make sure I'm completely clear...  You described various types of restraints, such as harmonic or flat bottomed.  That specifies how the energy varies with distance, but what precisely is the distance it depends on?  Is it the distance between the centers of mass of two groups of atoms?  Or something different?

**Comment by jchodera** [comment]
@davidlmobley: Standard Boresch-style restraints where each point used is a single atom are already possible.  We're just trying to establish if using sites defined by *groups* of atoms are important, and if so, what kinds of support is needed.

YANK uses harmonic restraints (though Boresch-style restraints can be added), but this is currently only limited to restraints between individual atoms.  @luirink would like to have a more general definition of the binding site that keeps the ligand near the center of the binding site---that requires defining mass-weighted centers of mass, centers of geometry, or some other weighted function of receptor atoms to define a site to be used in the restraints.  On the other hand, I'd just like to keep the ligand from floating too far from the protein, so some sort of minimum-distance restraint would be great.

**Comment by pren** [comment]
I suggest to stat with group based flat-bottomed harmonic restrain. If one atom is defined for each group, this reduces to atom based restraint. If the lower and upper restraint distances are set equal, it reduces to a simple harmonic potential. The distance can be defined as between the group center using user-specified weights as you suggested (mass, equal weights…)

**Comment by swails** [comment]
The forces being considered here are a lot more general than the flat-bottomed potential.  You already *have* atom-based flat-well potential restraints available through the use of `CustomBondForce`, `CustomAngleForce`, `CustomTorsionForce`, and `CustomCompoundBondForce` (that latter one can do a *lot* more).

A generalization of `CustomCompoundBondForce` to handle groups of atoms (rather than only individual atoms) will allow you to implement flat-well potentials, in addition to virtually every other functional type available -- including cylindrical and other types of restraints.

Given the power of the Lepton parser and custom energy expressions, there is virtually no advantage to implementing flat-well potentials specifically.  You might as well allow for an arbitrary energy expression.

**Comment by peastman** [comment]
Any preferences about what to call it?  CustomCenterOfMassForce?  CustomCentroidForce?  CustomGroupBondForce?

**Comment by rmcgibbo** [comment]
I like `CustomCentroidForce`, personally.

**Comment by swails** [comment]
`CustomCentroidCompoundBondForce`?  It's a little long, but potentially more descriptive.

**Comment by peastman** [comment]
Or `CustomCentroidBondForce`?  It's equally descriptive while being a little shorter.

**Comment by swails** [comment]
Sure.  I only suggested `CentroidCompoundBondForce` since I envisioned it mirroring the API and behavior of `CompoundBondForce`.  But `Compound` is unnecessary, really, as that functionality is a superset of a simple bond between centroids.

**Comment by jchodera** [comment]
If we're thinking of letting the centroids be weighted, `CustomCentroidBondForce` may not be fully descriptive, but it may be close enough.  Is `CustomWeightedCentroidBondForce` too silly?

**Comment by pren** [comment]
Is this limited to bond only? Or this is a generalization of CustomCompoundBondForce to groups?

**Comment by peastman** [comment]
My proposal above was that it be a generalization of CustomCompoundBondForce.

**Comment by pren** [comment]
Just replace existing CustomCompoundBondForce then? Also why call it BondForce? CustomCompoundForce?

**Comment by peastman** [comment]
No, this would be a new force.  It has a similar API, but is intended for different uses.  In principle the new class could be used to implement the same forces as CustomCompoundBondForce, but it would be both less convenient and less efficient.

**Comment by peastman** [comment]
Here's an initial version of the API: https://github.com/peastman/openmm/blob/centroid/openmmapi/include/openmm/CustomCentroidBondForce.h.  Tell me if you have any comments or suggestions.

**Comment by jchodera** [comment]
Awesome!  This seems to cover @luirink's use case nicely.

Some comments:
* What happens if fixed (zero-mass) atoms are included in a group?
* What happens if two groups contain one or more of the same atom(s)?
* Can groups contain single atoms?

**Comment by jlmaccal** [comment]
:+1: Looks great!

**Comment by swails** [comment]
> What happens if fixed (zero-mass) atoms are included in a group?

If you don't supply a weight, it's effectively excluded (there is no way to get around this).

> What happens if two groups contain one or more of the same atom(s)?

Good question.  I would vote on throwing an exception, since this is equivalent to simply doubling the weight.  There might be a use-case for keeping it, since you could *double* the weight of a particular atom without having to supply an explicit set of weights.  But my feeling is that a duplicated entry is more likely a mistake than intentional.

> Can groups contain single atoms?

I can't really imagine @peastman coding this in a way that you can't.  The centroid is still well-defined for a single atom.

Other situations to consider/catch (but these are implementation details more than API comments):

- If all of the weights are zero
- Negative weights (allowed? unallowed? can you do neat things with them?)

Overall on the API design: :+1:, I agree with @jlmaccal.

**Comment by peastman** [comment]
> What happens if fixed (zero-mass) atoms are included in a group?

That shouldn't be a problem.  Of course, if you use masses as weights, they won't actually contribute anything.

> What happens if two groups contain one or more of the same atom(s)?

I don't see any reason not to allow it.

> Can groups contain single atoms?

Of course.

> I would vote on throwing an exception, since this is equivalent to simply doubling the weight.

I think you're referring to something different: including an atom twice in a single group.  John was asking about including an atom in two different groups.

**Comment by swails** [comment]
> I think you're referring to something different: including an atom twice in a single group. John was asking about including an atom in two different groups.

I'm illiterate.

**Comment by peastman** [comment]
> Negative weights (allowed? unallowed? can you do neat things with them?)

Interesting question.  I can imagine cases where they'd be useful, so my inclination would be to allow them.

**Comment by swails** [comment]
Thinking about negative weights brings up another concern: PBCs.

How is that going to be handled here?

**Comment by pren** [comment]
If the restraint distance is less than half of the box, then minimum images should be fine. Otherwise it could be ambiguous.

PengYu Ren

**Comment by peastman** [comment]
That's an interesting question.  Here's how that's handled by CustomCompoundBondForce (and in fact all bonded forces):

1. Periodic boundary conditions are not applied when computing distances.

2. When translating atoms into a single periodic box, we always translate whole molecules (based on the positions of their centers).  If two atoms are in the same molecule, they always get translated together, so the distance between them is always well defined.

3. If two atoms are connected by any bonded force, then they are considered to be part of the same molecule.

Should CustomCentroidBondForce work in the same way?  Effectively that means it's an interaction between two *specific* groups of atoms.  Not between one group and whichever periodic copy of the other group happens to be closest, but between that group and a specific periodic copy of the other.

**Comment by jlmaccal** [comment]
I would say that, at least for the applications I can think of, I would want to respect PBC and use the minimum image convention. But there are certainly use cases for both ways.

How hard would it be to supply additional information (box vectors, etc) and necessary expressions to compute minimum image vectors as part of the expression parser?

**Comment by swails** [comment]
> Should CustomCentroidBondForce work in the same way?

Either that or, as @jlmaccal suggested, use the MIC (easy for distances; is it just as easy for angles and torsions?).  And then you have imaging *within* a defined group -- that would probably be easiest to treat as a single molecule.

It's hard to know what to vote for, but I think having `CustomCentroidBondForce` work the same way as `CustomCompoundBondForce` would be the most straightforward thing to do.  In my experience, umbrella sampling (and similar) setups typically follow the general workflow of:

1. Define the reaction coordinate you want to compute the PMF along.
2. Set up your system close to one end-point (e.g., a ligand bound in an active site)
3. Measure the value of the RC at that point, use that to set up the first "window" of umbrella sampling
4. Get snapshots along the RC, either using something like steered MD or a bunch of short US simulations starting from a snapshot from the previous window
5. Run all of your windows in parallel (doing the standard "heat, equilibrate, production" steps)

According to this workflow, equilibrium distances and angles are always set up according to measurements from a particular set of periodic images.  So treating connected centroids as "bonded" together (so that all atoms in all centroids are always part of the same molecule) will match the behavior that I'm used to and that works for the above workflow (I can't speak for everyone, though).  It will also avoid potential situations where a switch in which periodic image is used in the distance/angle/torsion calculation can blow up a simulation by introducing a massive discontinuity in the force.

Your proposed approach would lead, IMO, to the simplest design to understand and use.

Downsides include potential pressure artifacts if the centroids are *too big* and were never intended to be defined within a molecule-like collection of atoms.  The force considered in #1112 can be used to do stuff like restrain the closest water to an active site (by using a `min` reduce operation with the atom selections being "all waters" and "all residues in the active site") -- which would make everything part of the same molecule and make barostat moves no-ops.  I can't envision that kind of thing happening here, though.

**Comment by peastman** [comment]
Thinking about this further, I think it really ought to be treated as a bonded force, not a nonbonded one.  MIC is only appropriate for short range potentials that go to zero at some distance less than half the box size.  Otherwise, the force and energy change discontinuously when it jumps to a different periodic image.  The main use cases of this class are harmonic and flat bottomed harmonic potentials, which grow with distance rather than shrinking.

**Comment by peastman** [comment]
Implemented in #1119.
